src/main/scala/zio/s3/MultipartUploadOptions.scala

```diff
@@ -0,0 +1,35 @@
+package zio.s3
+
+import software.amazon.awssdk.services.s3.model.CreateMultipartUploadRequest
+import zio.s3.MultipartUploadOptions.MinMultipartPartSize
+
+import scala.jdk.CollectionConverters._
+
+/**
+ * The upload options that are specific to multipart uploads
+ *
+ * @param uploadOptions [[UploadOptions]]
+ * @param partSize the size of the part in bytes, the minimum is 5 MB
+ * @param parallelism the number of parallel requests to upload chunks, default to 1
+ */
+case class MultipartUploadOptions(
+  uploadOptions: UploadOptions = UploadOptions.default,
+  partSize: Int = MinMultipartPartSize,
```

Comment:
please put back the sealed trait

Reply:
Sorry in advance for the long read :smile: Here I am trying to explain why, I think, having no additional types other than Int would be my choice.

I thought that this comment was about not adding the sealed trait https://github.com/zio/zio-s3/pull/117#discussion_r487325347

In the example above that, you showed, there is a ChunkSize type
```
sealed trait ChunkSize {
  def size: Int
}
object ChunkSize {
  def from(size: Int): Either[String, ChunkSize] = ???
}
```
Most likely it is going to be used the way, that the chunk is validated an is immediately passed into the upload
```
  val result: ZIO[S3, io.Serializable, Unit] = for {
    partSize <- ZIO.fromEither(ChunkSize.from(100))
    _ <- multipartUpload("bucket", "key", ZStream.empty, MultipartUploadOptions(partSize = partSize))
  } yield ()
```
Here is the first problem: the error type is Serializable, because the multipartUpload's error is S3Exception, but the ChunkSize's error is a String.

Let's say we use the S3Exception in the "from" method (I also rename the ChunkSize to the PartSize, because that's what we use in the other names):
```
sealed trait PartSize {
  def size: Int
}
object PartSize {
  def from(size: Int): Either[S3Exception, PartSize] = ???
}
```
Now the error type is S3Exception everywhere.

But there is the problem that I mentioned earlier: the API is not consistent, because the "parallelism" is also a potentially invalid value but we pass it as a plain Int. This is what happens if we create Parallelism class and use it
```
  val result = for {
    partSize <- ZIO.fromEither(PartSize.from(100))
    parallelism <- ZIO.fromEither(Parallelism.from(4))
    _ <- multipartUpload("bucket", "key", ZStream.empty, MultipartUploadOptions(partSize = partSize))
  } yield ()
```
Everything compiles, the error is fine, we are sure that the values of parallelism and the partSize are valid. But the user is now responsible for validating all those values. If there is another option added that requires validation, the user would have to add another line with the ".from" statement.

If you really want to have the MultipartUploadOptions that have the fields as custom types rather than Ints, I would create the build for the options and make the case classe's constructor private like this
```
final case class PartSize(value: Int)    extends AnyVal
final case class Parallelism(value: Int) extends AnyVal

case class MultipartUploadOptions private (
  uploadOptions: UploadOptions,
  partSize: PartSize,
  parallelism: Parallelism
)

object MultipartUploadOptions {
  final val MinPartSize: Int = 5 * 1024 * 1024 // 5 MB

  def make(
    uploadOptions: UploadOptions = UploadOptions(),
    partSize: Int = MinPartSize,
    parallelism: Int = 1
  ): Either[S3Exception, MultipartUploadOptions] = ???
}
```

This way creating the options is always a single additional line, even if new options are added
```
  val result = for {
    options <- ZIO.fromEither(MultipartUploadOptions.make(partSize = 1000, parallelism = 4))
    _ <- multipartUpload("bucket", "key", ZStream.empty, options)
  } yield ()
```


Although having types instead of Ints sounds reasonable to me, as a user of the library I would prefer the library doing as much as possible for me. So it would be great if this additional line to create the options can be removed. After all, I will receive the S3Exception anyway. That's why I would prefer having Ints in the MultipartUploadOptions and don't the validation directly in "multipartUpload" method.